python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -0,0 +1,324 @@
+ */
+predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo) {
+  not nodeFrom.(EssaNode).getVar() instanceof GlobalSsaVariable and
+  not nodeTo.(EssaNode).getVar() instanceof GlobalSsaVariable and
+  EssaFlow::essaFlowStep(nodeFrom, nodeTo)
+}
+
+// TODO: Make modules for these headings
+//--------
+// Global flow
+//--------
+
+/** Represents a callable */
+class DataFlowCallable = CallableValue;
+
+/** Represents a call to a callable */
+class DataFlowCall extends CallNode {
+  DataFlowCallable callable;
+
+  DataFlowCall() {
+    this = callable.getACall()
+  }
+
+  /** Get the callable to which this call goes. */
+  DataFlowCallable getCallable() { result = callable }
+
+  /** Gets the enclosing callable of this call. */
+  DataFlowCallable getEnclosingCallable() {
+    result.getScope() = this.getNode().getScope()
+  }
+}
+
+/** A data flow node that represents a call argument. */
+class ArgumentNode extends CfgNode {
+  ArgumentNode() {
+    exists(DataFlowCall call, int pos |
+      node = call.getArg(pos)
+    )
+  }
+
+   /** Holds if this argument occurs at the given position in the given call. */
+  predicate argumentOf(DataFlowCall call, int pos) {
+    node = call.getArg(pos)
+  }
+
+   /** Gets the call in which this node is an argument. */
+  final DataFlowCall getCall() { this.argumentOf(result, _) }
+}
+
+/** Gets a viable run-time target for the call `call`. */
+DataFlowCallable viableCallable(DataFlowCall call) {
+  result = call.getCallable()
+}
+
+private newtype TReturnKind = TNormalReturnKind()
+
+/**
+ * A return kind. A return kind describes how a value can be returned
+ * from a callable. For Python, this is simply a method return.
+ */
+class ReturnKind extends TReturnKind {
+  /** Gets a textual representation of this return kind. */
+  string toString() { result = "return" }
+}
+
+/** A data flow node that represents a value returned by a callable. */
+class ReturnNode extends CfgNode {
+  Return ret;
+
+   // See `TaintTrackingImplementation::returnFlowStep`
+  ReturnNode() {
+    node = ret.getValue().getAFlowNode()
+  }
+
+   /** Gets the kind of this return node. */
+  ReturnKind getKind() { result = TNormalReturnKind() }
+
+   override DataFlowCallable getEnclosingCallable() {
+    result.getScope().getAStmt() = ret // TODO: check nested function definitions
+  }
+}
+
+/** A data flow node that represents the output of a call. */
+class OutNode extends CfgNode {
+  OutNode() { node instanceof CallNode }
+
+   /** Gets the underlying call, where this node is a corresponding output of kind `kind`. */
+  cached
+  DataFlowCall getCall(ReturnKind kind) {
+    kind = TNormalReturnKind() and
+    result = node
+  }
+}
+
+/**
+ * Gets a node that can read the value returned from `call` with return kind
+ * `kind`.
+ */
+OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) { call = result.getCall(kind) }
+
+//--------
+// Type pruning
+//--------
+
+newtype TDataFlowType =
+  TStringFlow()
```

Comment:
This seems to be used as a dummy type right now. Can we use a more appropriate name than String? :sweat_smile:

Reply:
I changed it to `TAnyFlow`